tensorizer/serialization.py

```diff
@@ -912,6 +997,32 @@ def load_into_module(
         self._file.close()
         return tensor_ct
 
+    def verify_module(self,
+                      m: torch.nn.Module) -> None:
+        """
+        Given `m`, a torch.nn.Module, verify that the tensors in this
+        Tensorizer object match the tensors in the `torch.nn.Module`.
+        Mismatches will raise a ValueError.
+        """
+        modules: typing.OrderedDict[str, torch.nn.Module] = OrderedDict()
+
+        for name, module in m.named_modules():
+            modules[name] = module
+
+        for name in self.keys():
+            obj_path, attr = name.rsplit(".", 1)
+            module: torch.nn.Module = modules[obj_path]
+            entry = self._metadata[name]
+            # Check if the module has the attribute
+            if not hasattr(module, attr):
+                raise ValueError(
+                    f"Module {obj_path} does not have attribute {attr}"
+                )
+            numpy_tensor = _NumpyTensor.from_tensor(getattr(module, attr))
+            self._verify_hashes(name,
+                                entry["hashes"],
+                                entry["raw_headers"],
+                                numpy_tensor.data.tobytes())
```

Comment:
Two concerns:
1. This does a one-way comparison, so it passes as long as the `TensorDeserializer` is a subset of `m`.
    1. This means even an empty `TensorDeserializer` would pass this check
    2. This needs a supplementary check that `self.keys() == m_keys` where `m_keys = m.state_dict().keys() | dict(m.named_buffers()).keys()`
    3. Like for `check_deserialized` in `tests/test_serialization.py`, this gets ambiguous when the `TensorDeserializer` is filtered by a `filter_func`, which can be solved in two ways:
        1. Add an `allow_subset` flag that uses the current behaviour, *and/or*
        2. Add a `filter_func` parameter to `verify_module` that lets someone choose to only compare against a specific subset of `m_keys`
2. A function that has no observable effect or return value but sometimes raises an exception seems less useful than one that returns a `bool`, since with a bool you can choose to e.g. `assert verify_module(...)` or use it in an if-statement.

Reply:
1. I don't think we can easily compare tensors for the absence of a tensor in the module, as listing the named modules yields things that don't quite 100% match. I'll need to noodle on this. We'll add a `TODO` for this.
2. The function has been revamped to yield a boolean and a list of tuples of name, bool indicating the status of the individual tensors. A test case has been added for a tensor that has changed.

Reply:
[`check_deserialized` in the test suite](https://github.com/coreweave/tensorizer/blob/0478eed41fa14a874ac82bcaa06adf4def2eca24/tests/test_serialization.py#L41-L48) implements such a comparison between the keys in a deserialized module versus a reference module. It should be usable as a base for `verify_module` as well:

```python
def check_deserialized(deserialized, model_name: str, allow_subset=False):
    orig_model = AutoModelForCausalLM.from_pretrained(model_name)
    orig_sd = orig_model.state_dict()
    # Non-persistent buffers are serialized in tensorizer,
    # but aren't included in a state_dict() in PyTorch.
    orig_sd.update(orig_model.named_buffers())
    if not allow_subset:
        assert orig_sd.keys() == deserialized.keys()
    ...
```

`state_dict() | named_buffers()` (and potentially some other unions) seem to give a comprehensive set of entries, but it'd need to be tested on more models than `gpt-neo-125M` to make sure of that.

Reply:
Hm. What's the overhead of doing that, though?